### PR TITLE
Deterministic filter-like device

### PR DESCRIPTION
**Pull request by richfitz**
~Merge after #2, contains those commits~
~Merge after #11, contains those commits~

This PR implements a particle-filter-like method for computing likelihoods out of a time series model by comparing to data periodically.  It is the limit case of a particle filter to one particle, and has been used as the "deterministic" method in sircovid/dust1/etc.

The name `unfilter` is temporary, we'll probably come up with something better later.

The core implementation is in `inst/dust2/filter.hpp` - you'll see it's pretty straightforward - we loop over all times in the data set, do the comparison to data and accumulate a likelihood.  The R wrapper around this then returns that likelihood out of the container back to R.

We'll complicate this shortly with things like returning trajectories as the model runs.

The time handling, and in particular the conversion to `steps_` is pretty gross and addressed somewhat in #12, which we might use here soon

The current version only allows using initial conditions from the model, this is addressed in #13

### CODE REVIEW

File: tests/testthat/test-filter.R [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    for (i in seq_along(time)) {
+      dust2_cpu_sir_run_steps(ptr, round((time[i] - time0[i]) / dt))
+      incidence[i] <- dust2_cpu_sir_state(ptr, FALSE)[5, , drop = TRUE]
+    }
+    sum(dpois(1:4, incidence + 1e-6, log = TRUE))
+  }
+
+  obj <- dust2_cpu_sir_unfilter_alloc(base, time_start, time, dt, data, 1, 0)
+  ptr <- obj[[1]]
+  expect_equal(dust2_cpu_sir_unfilter_run(ptr, NULL, FALSE), f(pars1))
```
Comment (weshinsley) [inline tests/testthat/test-filter.R:28]: So this one with `NULL, FALSE` uses `base` - which happens to be the same as pars1. We could have said `f(base)` on the end.

File: tests/testthat/test-filter.R [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      dust2_cpu_sir_run_steps(ptr, round((time[i] - time0[i]) / dt))
+      incidence[, i] <- dust2_cpu_sir_state(ptr, FALSE)[5, , drop = TRUE]
+    }
+    observed <- matrix(unlist(data, use.names = FALSE), n_groups)
+    rowSums(dpois(observed, incidence + 1e-6, log = TRUE))
+  }
+
+  obj <- dust2_cpu_sir_unfilter_alloc(pars, time_start, time, dt, data, 1, 3)
+  ptr <- obj[[1]]
+  expect_equal(dust2_cpu_sir_unfilter_run(ptr, NULL, TRUE), f(pars))
```
Comment (weshinsley) [inline tests/testthat/test-filter.R:65]: Here, pars[[1 .. 3]] is an entry for each group, and f(pars) gives us 3 results, so a summed-log-likelihood over time, for each of the 3 groups.

File: tests/testthat/test-filter.R [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  obj <- dust2_cpu_sir_unfilter_alloc(base, time_start, time, dt, data, 1, 0)
+  ptr <- obj[[1]]
+  expect_equal(dust2_cpu_sir_unfilter_run(ptr, NULL, FALSE), f(pars1))
+
+  expect_equal(dust2_cpu_sir_unfilter_run(ptr, pars1, FALSE), f(pars1))
+  expect_equal(dust2_cpu_sir_unfilter_run(ptr, pars2, FALSE), f(pars2))
+})
+
+
+test_that("can run run unfilter on structured model", {
```
Comment (weshinsley) [inline tests/testthat/test-filter.R:35]:
```suggestion
test_that("can run unfilter on structured model", {
```

File: tests/testthat/test-filter.R [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  base <- list(beta = 0.1, gamma = 0.2, N = 1000, I0 = 10, exp_noise = 1e6)
+  n_groups <- 3
+  pars <- lapply(seq_len(n_groups),
+                 function(i) modifyList(base, list(beta = i * 0.1)))
+
+  time_start <- 0
+  time <- c(4, 8, 12, 16)
+  data <- lapply(1:4, function(i) {
+    lapply(seq_len(n_groups), function(j) list(incidence = 2 * (i - 1) + j))
+  })
```
Comment (weshinsley) [inline tests/testthat/test-filter.R:45]: So data[[1 .. 4]] each contain 3 incidence points, one for each group.

Comment (weshinsley) [reply tests/testthat/test-filter.R:45]: and the 1 .. 4 in data refer to the 4 timesteps

File: tests/testthat/test-filter.R [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+test_that("validate time for filter", {
+  pars <- list(beta = 0.1, gamma = 0.2, N = 1000, I0 = 10, exp_noise = 1e6)
+  time_start <- 0
+  time <- as.integer(c(4, 8, 12, 16))
+  data <- lapply(1:4, function(i) list(incidence = i))
+  dt <- 1
+
+  expect_error(
+    dust2_cpu_sir_unfilter_alloc(pars, 5, time, dt, data, 1, 0),
```
Comment (weshinsley) [inline tests/testthat/test-filter.R:77]: 2nd arg (5) is r_time_start.
We expect (and get) errors if r_time_start >= length(time)

**Review by weshinsley — COMMENTED** [review]
Good - think I undesrstand this - one typo I'll commit, and some of my train of thought running through the tests.

**Review by weshinsley — APPROVED** [review]

File: tests/testthat/test-filter.R (hunk 1 above)
Comment (weshinsley) [reply tests/testthat/test-filter.R:28]: General comment - I find it quite easy to forget what the args are with the functions...!

Comment (richfitz) [reply tests/testthat/test-filter.R:28]: yeah, I'll sort this out in one push once we get an actual interface to any of this stuff - none of these functions will be the ones we call in the end!

**Pull request merged**